src/internal/communication.ts

```diff
@@ -0,0 +1,472 @@
+    Communication.handlers['closeConversation'] = Communication.handleCloseConversation;
+    Communication.handlers['appButtonClick'] = Communication.handleAppButtonClick;
+    Communication.handlers['appButtonHoverEnter'] = Communication.handleAppButtonHoverEnter;
+    Communication.handlers['appButtonHoverLeave'] = Communication.handleAppButtonHoverLeave;
+
+    // Listen for messages post to our window
+    Communication.messageListener = (evt: DOMMessageEvent): void => Communication.processMessage(evt);
+
+    // If we are in an iframe, our parent window is the one hosting us (i.e., window.parent); otherwise,
+    // it's the window that opened us (i.e., window.opener)
+    Communication.currentWindow = Communication.currentWindow || window;
+    Communication.parentWindow =
+      Communication.currentWindow.parent !== Communication.currentWindow.self
+        ? Communication.currentWindow.parent
+        : Communication.currentWindow.opener;
+
+    // Listen to messages from the parent or child frame.
+    // Frameless windows will only receive this event from child frames and if validMessageOrigins is passed.
+    if (Communication.parentWindow || validMessageOrigins) {
+      Communication.currentWindow.addEventListener('message', Communication.messageListener, false);
+    }
+
+    if (!Communication.parentWindow) {
+      GlobalVars.isFramelessWindow = true;
+      // @ts-ignore: window as ExtendedWindow
+      (window as ExtendedWindow).onNativeMessage = Communication.handleParentMessage;
+    }
+
+    try {
+      // Send the initialized message to any origin, because at this point we most likely don't know the origin
+      // of the parent window, and this message contains no data that could pose a security risk.
+      Communication.parentOrigin = '*';
+      Communication.sendMessageToParent('initialize', [version], callback);
+    } finally {
+      Communication.parentOrigin = null;
+    }
+  }
+
+  public static uninitialize(): void {
+    Communication.currentWindow.removeEventListener('message', Communication.messageListener, false);
+
+    Communication.parentWindow = null;
+    Communication.parentOrigin = null;
+    Communication.parentMessageQueue = [];
+    Communication.childWindow = null;
+    Communication.childOrigin = null;
+    Communication.childMessageQueue = [];
+    Communication.nextMessageId = 0;
+    Communication.callbacks = {};
+  }
+
+  /**
+   * Send a message to parent. Uses nativeInterface on mobile to communicate with parent context
+   */
+  public static sendMessageToParent(actionName: string): void;
+  /**
+   * Send a message to parent. Uses nativeInterface on mobile to communicate with parent context
+   */
+  public static sendMessageToParent(actionName: string, args: any[]): void;
+  /**
+   * Send a message to parent. Uses nativeInterface on mobile to communicate with parent context
+   */
+  public static sendMessageToParent(actionName: string, callback: Function): void;
```

Comment:
is there any better way to minimum no of declaration for sendMessageToParent function?

Reply:
Good point! I have reduced it to 2